### PR TITLE
feat: Added reward model according to paper.

### PR DESCRIPTION
**Pull request by ahmeterdempmk**
This PR implements the reward model training pipeline for DeepSeek-R1, adding key functionality for preference learning and model comparison.

## Changes
- Added`reward_model.py` implementing reward model training pipeline
- Introduced`RewardModelScriptArguments` class for handling training configuration
- Implemented dataset preparation with`prepare_comparison_dataset` function
- Added support for better/worse response comparison training
- Integrated with HuggingFace's Trainer for model training
- Added model saving and Hub pushing capabilities

## Implementation Details
- Uses AutoModelForSequenceClassification for reward modeling
- Supports configurable dataset columns for comparisons
- Implements efficient batch processing with proper tokenization
- Handles model checkpointing and evaluation

## Testing
The implementation has been tested with:

- Dataset preparation and tokenization
- Model training pipeline
- Hub integration for model sharing

## Next Steps
- Add evaluation metrics for reward model performance
- Implement additional data preprocessing options
- Add documentation for training configuration

### CODE REVIEW

File: src/open_r1/reward_model.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        num_labels=1,
+        trust_remote_code=model_args.trust_remote_code
+    )
+    
+    # Load and prepare dataset
+    dataset = load_dataset(script_args.dataset_name, name=script_args.dataset_config)
+    processed_dataset = prepare_comparison_dataset(dataset, tokenizer, script_args)
+    
+    # Initialize trainer
+    trainer = Trainer(
```
Comment (qgallouedec) [inline src/open_r1/reward_model.py:95]: Thanks! Any reason for not using the RewardTrainer?

Comment (ahmeterdempmk) [reply src/open_r1/reward_model.py:95]: Because the implementation focuses on direct preference comparison training with better/worse response pairs, which is effectively handled by the standard Trainer with our custom dataset preparation.

Comment (qgallouedec) [reply src/open_r1/reward_model.py:95]: Okay, but that's exactly what the RewardTrainer does, isn't it?

**Comment by qgallouedec** [comment]
Can you also explain which part of the paper this addition corresponds to? It seems to me that the idea behind the RL of DeepSeek-R1 is to free it from reward models.
